domain/src/test/java/org/oppia/domain/topic/StoryProgressControllerTest.kt

```diff
@@ -136,116 +179,249 @@ class StoryProgressControllerTest {
+      assertThat(recordProgressResultCaptor.value.isSuccess()).isTrue()
+
+      verify(mockStoryProgressObserver, atLeastOnce()).onChanged(storyProgressResultCaptor.capture())
+      assertThat(storyProgressResultCaptor.value.isSuccess()).isFalse()
+    }
 
   @Test
-  fun testRecordCompletedChapter_validSecondStory_validChapter_notYetCompleted_marksChapterAsCompleted() {
-    storyProgressController.recordCompletedChapter(TEST_STORY_ID_2, TEST_EXPLORATION_ID_4)
+  @ExperimentalCoroutinesApi
+  fun testRecordCompletedChapter_validData_recordProgress_getChapterProgress_chapterProgressIsCorrect() =
+    runBlockingTest(coroutineContext) {
+      storyProgressController.recordCompletedChapter(PROFILE_ID_1, TOPIC_ID_1, STORY_ID_1, EXPLORATION_ID_1)
+        .observeForever(mockRecordProgressObserver)
+      advanceUntilIdle()
 
-    val storyProgress = storyProgressController.getStoryProgress(TEST_STORY_ID_2).value!!.getOrThrow()
-    assertThat(storyProgress.getChapterProgress(0).explorationId).isEqualTo(TEST_EXPLORATION_ID_4)
-    assertThat(storyProgress.getChapterProgress(0).playState).isEqualTo(COMPLETED)
-  }
+      storyProgressController.getChapterProgress(PROFILE_ID_1, TOPIC_ID_1, STORY_ID_1, EXPLORATION_ID_1)
+        .observeForever(mockChapterProgressObserver)
+
+      verify(mockRecordProgressObserver, atLeastOnce()).onChanged(recordProgressResultCaptor.capture())
+      assertThat(recordProgressResultCaptor.value.isSuccess()).isTrue()
+
+      verify(mockChapterProgressObserver, atLeastOnce()).onChanged(chapterProgressResultCaptor.capture())
+      assertThat(chapterProgressResultCaptor.value.isSuccess()).isTrue()
+
+      val chapterProgress = chapterProgressResultCaptor.value.getOrThrow()
+      assertThat(chapterProgress).isNotNull()
+      assertThat(chapterProgress).isEqualTo(COMPLETED)
+    }
 
   @Test
-  fun testRecordCompletedChapter_validSecondStory_validChapterInOtherStory_fails() {
-    val recordProgressLiveData = storyProgressController.recordCompletedChapter(TEST_STORY_ID_2, TEST_EXPLORATION_ID_3)
+  @ExperimentalCoroutinesApi
+  fun testRecordCompletedChapter_validData_recordProgress_getChapterProgressForInvalidProfileID_noResultFound() =
+    runBlockingTest(coroutineContext) {
+      storyProgressController.recordCompletedChapter(PROFILE_ID_1, TOPIC_ID_1, STORY_ID_1, EXPLORATION_ID_1)
+        .observeForever(mockRecordProgressObserver)
+      advanceUntilIdle()
 
-    val recordProgressResult = recordProgressLiveData.value
-    assertThat(recordProgressResult).isNotNull()
-    assertThat(recordProgressResult!!.isFailure()).isTrue()
-    assertThat(recordProgressResult.getErrorOrNull())
-      .hasMessageThat()
-      .contains("Chapter not found in story: $TEST_EXPLORATION_ID_3")
-  }
+      storyProgressController.getChapterProgress(PROFILE_ID_2, TOPIC_ID_1, STORY_ID_1, EXPLORATION_ID_1)
+        .observeForever(mockChapterProgressObserver)
```

Comment:
Should there be an ``advanceUntilIdle`` call after the above? At first glance, it sort of seems like that could cause a failure rather than the underlying state being set up (specifically because it's inconsistent with the previous ``recordCompletedChapter`` call).